File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        
        vector<string> res;
        
        string current;
        backtrack(res, current, 0, 0, n);
        return res;
    }
    
private:
    void backtrack(vector<string> &res, string current, int openCount, int closeCount, int n){
        if(openCount == closeCount && openCount == n){
            res.push_back(current);
            return;
        }
        
        if(openCount < n){
            current += "(";
            backtrack(res, current, openCount + 1, closeCount, n);
            current.pop_back();
        }
        
        if(closeCount < openCount){
            current += ")";
            backtrack(res, current, openCount, closeCount + 1, n);
            current.pop_back();
        }
    }
};

/*
int main() {
    Solution solution;
    int n = 3; // You can change this value to test with different n
    vector<string> result = solution.generateParenthesis(n);

    // Print the generated parenthesis combinations
    for (const string& combination : result) {
        cout << combination << endl;
    }

    return 0;
}
*/

File: 0022-generate-parentheses/0022-generate-parentheses.py
## space : o(n)
## time : o(2^n)

class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        
        ## only add open parenthesis if opencount < n
        ## only add close parenthesis if closecount < opencount
        ## valid if open == closed == n
        
        ## initialize an empty stack
        
        stack = []
        
        res = []
        
        ## recursive approach
        def backtrack(openCount, closeCount):
            
            ##check if both open and close counts are equal
            
            if(openCount == closeCount == n):
                
                ## from that stack convert the parenthesis into string and add into result array
                
                res.append("".join(stack))
                return
            
            if(openCount < n):
                
                
                ## add open parenthesis to the stack
                stack.append("(")
                ##recursive call
                backtrack(openCount + 1, closeCount)
                stack.pop()
                
            if(closeCount < openCount):
                ## add close parentthesis to the staxk
                stack.append(")")
                backtrack(openCount, closeCount + 1)
                stack.pop()
                
        ## initialize the backtrack function with open and closose count as 0, 0
        backtrack(0, 0)
        return res
                
        